### PR TITLE
[Merged by Bors] - feat(Data/List): Some auxiliary lemmas of `List`

### PR DESCRIPTION
**Pull request by znssong**
Add some auxiliary lemmas of `List`, in preparation of the already finished proof of Bondy-Chvátal theorem.

---
- [x] depends on: #15876

It was discussed on the thread [](https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there-code-for-X.3F/topic/Brooks'.20theorem).

### CODE REVIEW

File: Mathlib/Data/List/Lemmas.lean [hunk 1]
```diff
@@ -18,6 +18,43 @@ variable {α β γ : Type*}
...
+  omega
+
+theorem head_reverse {α} (l : List α) (hL : l ≠ []) :
+    l.reverse.head (by simpa) = l.getLast hL := by
+  have : l.reverse.head? = l.getLast? := by simp
+  rw [List.head?_eq_head (l := l.reverse) (by simpa)] at this
+  rw [List.getLast?_eq_getLast (l := l) hL] at this
+  exact Option.some_inj.mp this
+
+lemma tail_get {α} {i} (L : List α) (hi : i < L.tail.length) :
```
Comment (eric-wieser) [inline Mathlib/Data/List/Lemmas.lean:54]:
```suggestion
lemma getElem_tail {α} {i} (L : List α) (hi : i < L.tail.length) :
```
I think is the usual naming here

File: Mathlib/Data/List/Rotate.lean [hunk 2]
```diff
@@ -477,6 +477,25 @@ theorem IsRotated.map {β : Type*} {l₁ l₂ : List α} (h : l₁ ~r l₂) (f :
...
   use n
 
+theorem IsRotated.cons_getLast_dropLast
+    {α} (L : List α) (hL : L ≠ []) : (L.getLast hL :: L.dropLast) ~r L := by
+  induction L using List.reverseRecOn with
+  | nil => simp at hL
+  | append_singleton a L _ =>
+    simp
+    apply List.IsRotated.symm
+    apply List.isRotated_concat
```
Comment (qawbecrdtey) [inline Mathlib/Data/List/Rotate.lean:487]:
```suggestion
    apply IsRotated.symm
    apply isRotated_concat
```

File: Mathlib/Data/List/Rotate.lean [hunk 3]
```diff
@@ -477,6 +477,25 @@ theorem IsRotated.map {β : Type*} {l₁ l₂ : List α} (h : l₁ ~r l₂) (f :
...
+
+theorem IsRotated.dropLast_tail {α}
+    {L : List α} (hL : L ≠ []) (hL' : L.head hL = L.getLast hL) : L.dropLast ~r L.tail :=
+  match L with
+  | [] => by simp
+  | [_] => by simp
+  | a :: b :: L => by
+    simp at hL' |-
+    rw [hL']
+    apply IsRotated.cons_getLast_dropLast
```
Comment (qawbecrdtey) [inline Mathlib/Data/List/Rotate.lean:497]:
```suggestion
    simp [hL', IsRotated.cons_getLast_dropLast]
```

**Review by qawbecrdtey — COMMENTED** [review]
Maybe you should try merging the master branch?

File: Mathlib/Data/List/Lemmas.lean [hunk 4]
```diff
@@ -18,6 +18,43 @@ variable {α β γ : Type*}
 
 namespace List
 
+theorem getElem_reverse' {α} (l : List α) (i : Nat) (h1 h2) :
+    (reverse l)[i]'h1 = l[length l - 1 - i]'h2 := by
+  have := getElem_reverse l (length l - 1 -i) (by omega) (by omega)
+  rw [← this]
```
Comment (eric-wieser) [inline Mathlib/Data/List/Lemmas.lean:24]: nit:
```suggestion
  rw [← getElem_reverse l _ (by omega) (by omega)]
```

File: Mathlib/Data/List/Lemmas.lean [hunk 5]
```diff
@@ -18,6 +18,43 @@ variable {α β γ : Type*}
 
 namespace List
 
+theorem getElem_reverse' {α} (l : List α) (i : Nat) (h1 h2) :
```
Comment (eric-wieser) [inline Mathlib/Data/List/Lemmas.lean:21]: There is already an ambient `α` in the file
```suggestion
theorem getElem_reverse' (l : List α) (i : Nat) (h1 h2) :
```
and throughout the diff.

File: Mathlib/Data/List/Lemmas.lean [hunk 6]
```diff
@@ -18,6 +18,43 @@ variable {α β γ : Type*}
...
+    rw [getElem?_eq_getElem (by simpa), getElem?_eq_getElem (by simpa),
+      ← getElem_drop _, getElem_reverse', getElem_reverse', getElem_dropLast]
+    simp [show l.length - 1 - (1 + i) = l.length - 1 - 1 - i by omega]
+    all_goals ((try simp); omega)
+  · rw [getElem?_eq_none, getElem?_eq_none]
+    all_goals (simp; omega)
+
+theorem getLast_tail {α} (l : List α) (hl : l.tail ≠ []) :
+    l.tail.getLast hl = l.getLast (by intro h; rw [h] at hl; simp at hl) := by
+  simp [getLast_eq_get, ← drop_one] at hl |-
```
[the remark this answers is not in the document]
Comment (meow-sister) [reply Mathlib/Data/List/Lemmas.lean:41]: OK, thanks.

[the remark this answers is not in the document]
Comment (meow-sister) [reply Mathlib/Data/List/Lemmas.lean:41]: Which tactic sequence will be considered as "non-terminal `simp`"? Does `simp; omega` also non-terminal?

[the remark this answers is not in the document]
Comment (meow-sister) [reply Mathlib/Data/List/Lemmas.lean:41]: Or `simp at hl; simp`

[the remark this answers is not in the document]
Comment (Ruben-VandeVelde) [reply Mathlib/Data/List/Lemmas.lean]: `simp; omega` is probably fine. `simp at hl; simp [hl]` is not.

File: Mathlib/Data/List/Lemmas.lean [hunk 7]
```diff
@@ -18,13 +18,41 @@ variable {α β γ : Type*}
 
 namespace List
 
+theorem getElem_reverse' (l : List α) (i : Nat) (h1 h2) :
```
Comment (YaelDillies) [inline Mathlib/Data/List/Lemmas.lean:21]:
```suggestion
-- TODO: Replace `List.getElem_reverse`
theorem getElem_reverse' (l : List α) (i : Nat) (h1 h2) :
```
as it seems like a strictly better lemma

File: Mathlib/Data/List/Lemmas.lean [hunk 8]
```diff
@@ -18,13 +18,41 @@ variable {α β γ : Type*}
...
+    (reverse l)[i]'h1 = l[length l - 1 - i]'h2 := by
+  rw [← getElem_reverse l _ (by omega) (by omega)]
+  congr
+  simp at h1
+  omega
+
+theorem reverse_tail_eq_dropLast_reverse {α} (l : List α) :
+    l.reverse.tail = l.dropLast.reverse := by
```
Comment (YaelDillies) [inline Mathlib/Data/List/Lemmas.lean:29]:
```suggestion
theorem tail_reverse_eq_reverse_dropLast (l : List α) :
    l.reverse.tail = l.dropLast.reverse := by
```

File: Mathlib/Data/List/Lemmas.lean [hunk 9]
```diff
@@ -18,13 +18,41 @@ variable {α β γ : Type*}
...
+
+theorem getLast_tail {α} (l : List α) (hl : l.tail ≠ []) :
+    l.tail.getLast hl = l.getLast (by intro h; rw [h] at hl; simp at hl) := by
+  simp only [← drop_one, ne_eq, drop_eq_nil_iff_le,
+    not_le, getLast_eq_getElem, length_drop] at hl |-
+  rw [← getElem_drop]
+  simp [show 1 + (l.length - 1 - 1) = l.length - 1 by omega]
+  omega
+
+lemma getElem_tail {α} {i} (L : List α) (hi : i < L.tail.length) :
```
Comment (YaelDillies) [inline Mathlib/Data/List/Lemmas.lean:47]:
```suggestion
lemma getElem_tail {i} (L : List α) (hi : i < L.tail.length) :
```

File: Mathlib/Data/List/Rotate.lean [hunk 10]
```diff
@@ -477,6 +477,24 @@ theorem IsRotated.map {β : Type*} {l₁ l₂ : List α} (h : l₁ ~r l₂) (f :
   rw [map_rotate]
   use n
 
+theorem IsRotated.cons_getLast_dropLast
+    {α} (L : List α) (hL : L ≠ []) : (L.getLast hL :: L.dropLast) ~r L := by
```
Comment (YaelDillies) [inline Mathlib/Data/List/Rotate.lean:481]:
```suggestion
    (L : List α) (hL : L ≠ []) : L.getLast hL :: L.dropLast ~r L := by
```

File: Mathlib/Data/List/Lemmas.lean [hunk 11]
```diff
@@ -18,13 +18,42 @@ variable {α β γ : Type*}
...
+  ext i v; by_cases hi : i < l.length - 1
+  · simp only [← drop_one]
+    rw [getElem?_eq_getElem (by simpa), getElem?_eq_getElem (by simpa),
+      ← getElem_drop _, getElem_reverse', getElem_reverse', getElem_dropLast]
+    simp [show l.length - 1 - (1 + i) = l.length - 1 - 1 - i by omega]
+    all_goals ((try simp); omega)
+  · rw [getElem?_eq_none, getElem?_eq_none]
+    all_goals (simp; omega)
+
+theorem getLast_tail {α} (l : List α) (hl : l.tail ≠ []) :
```
Comment (YaelDillies) [inline Mathlib/Data/List/Lemmas.lean:40]:
```suggestion
theorem getLast_tail (l : List α) (hl : l.tail ≠ []) :
```
and below

File: Mathlib/Data/List/Rotate.lean [hunk 12]
```diff
@@ -477,6 +477,24 @@ theorem IsRotated.map {β : Type*} {l₁ l₂ : List α} (h : l₁ ~r l₂) (f :
   rw [map_rotate]
   use n
 
+theorem IsRotated.cons_getLast_dropLast
+    (L : List α) (hL : L ≠ []) : (L.getLast hL :: L.dropLast) ~r L := by
```
Comment (YaelDillies) [inline Mathlib/Data/List/Rotate.lean:481]:
```suggestion
    (L : List α) (hL : L ≠ []) : L.getLast hL :: L.dropLast ~r L := by
```
If it doesn't work, please tell me and we will fix the notation precedences

Comment (meow-sister) [reply Mathlib/Data/List/Rotate.lean]: I'm sorry, but it really fails.

Comment (YaelDillies) [reply Mathlib/Data/List/Rotate.lean]: Should be fixed after #15876

Comment (YaelDillies) [reply Mathlib/Data/List/Rotate.lean]: Can you please merge master now?

Comment (meow-sister) [reply Mathlib/Data/List/Rotate.lean]: Merged. Sorry for reply lately...

**Comment by YaelDillies** [comment]
Thanks!

maintainer merge

**Comment by mattrobball** [comment]
bors merge

**Pull request closed without merging**